### PR TITLE
Migrate no-string-literal rule to 1.4

### PR DESCRIPTION
**Pull request by adidahiya**

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -159,6 +160,7 @@ module.exports = function(grunt) {
           'test/rules/noDuplicateVariableRuleTests.ts',
           'test/rules/noEmptyRuleTests.ts',
           'test/rules/noEvalRuleTests.ts',
+          './test/rules/noStringLiteralRuleTests.ts',
```
Comment (ashwinr) [inline Gruntfile.js]: no leading period

File: src/rules/noStringLiteralRule.ts [hunk 2]
```diff
@@ -19,34 +19,36 @@
...
-    public visitElementAccessExpression(node: TypeScript.ElementAccessExpressionSyntax): void {
-        this.handleElementAccessExpression(node);
-        super.visitElementAccessExpression(node);
-    }
-
-    private handleElementAccessExpression(node: TypeScript.ElementAccessExpressionSyntax) {
+    public visitElementAccessExpression(node: ts.ElementAccessExpression) {
         var argument = node.argumentExpression;
-        var id = TypeScript.fullText(argument);
+        var accessor = argument.getText();
```
Comment (ashwinr) [inline src/rules/noStringLiteralRule.ts]: keep names consistent. call this text since this is the return from getText()

Comment (ashwinr) [inline src/rules/noStringLiteralRule.ts]: oh, i see why you're calling it accessor. accessorText is more appropriate here.

File: src/rules/noStringLiteralRule.ts [hunk 3]
```diff
@@ -19,34 +19,36 @@
...
 
         // the argument expression should be a string of length at least 2 (due to quote characters)
-        if (argument.kind() !== TypeScript.SyntaxKind.StringLiteral || id.length < 2) {
-            return;
+        if (argument.kind === ts.SyntaxKind.StringLiteral && accessor.length > 2) {
+            var unquotedAccessor = accessor.substring(1, accessor.length - 1);
```
Comment (ashwinr) [inline src/rules/noStringLiteralRule.ts]: unquotedAccessorText

File: test/helper.ts [hunk 4]
```diff
@@ -64,23 +64,17 @@ module Lint.Test {
...
+        });
     }
 
     // assert whether a failure array contains the given failure
     export function assertContainsFailure(haystack: Lint.RuleFailure[], needle: Lint.RuleFailure) {
-        for (var i = 0; i < haystack.length; ++i) {
-            if (haystack[i].equals(needle)) {
-                return;
-            }
+        if (haystack.filter((failure) => failure.equals(needle)).length === 0) {
```
Comment (ashwinr) [inline test/helper.ts]: move this out of the if

File: test/rules/noStringLiteralRuleTests.ts [hunk 5]
```diff
@@ -20,11 +20,29 @@ describe("<no-string-literal>", () => {
     it("forbids object access via string literals", () => {
         var fileName = "rules/sub.test.ts";
         var NoStringLiteralRule = Lint.Test.getRule("no-string-literal");
-        var failureString = NoStringLiteralRule.FAILURE_STRING;
 
         var actualFailures = Lint.Test.applyRuleOnFile(fileName, NoStringLiteralRule);
-        var expectedFailure = Lint.Test.createFailure(fileName, [11, 21], [11, 24], failureString);
 
-        Lint.Test.assertContainsFailure(actualFailures, expectedFailure);
+        var expectedFailure1 = Lint.Test.createFailure(fileName,
```
Comment (ashwinr) [inline test/rules/noStringLiteralRuleTests.ts]: just write a single line when you're creating these failures.

File: src/rules/noStringLiteralRule.ts [hunk 6]
```diff
@@ -19,34 +19,36 @@
...
+        super.visitElementAccessExpression(node);
+    }
 
-        // only create a failure if the identifier is valid, in which case there's no need to use string literals
-        if (isValidIdentifier) {
-            var position = this.positionAfter(node.expression, node.openBracketToken);
-            this.addFailure(this.createFailure(position, TypeScript.width(argument), Rule.FAILURE_STRING));
-        }
+    private handleElementAccessExpression(node: ts.ElementAccessExpression) {
+    }
```
Comment (adidahiya) [inline src/rules/noStringLiteralRule.ts]: empty block. awkward

Comment (ashwinr) [inline src/rules/noStringLiteralRule.ts]: remove it.

File: test/rules/noStringLiteralRuleTests.ts (hunk 5 above)
Comment (ashwinr) [inline test/rules/noStringLiteralRuleTests.ts]: single lines here

Comment (adidahiya) [inline test/rules/noStringLiteralRuleTests.ts]: nah, i like this, I did it for typedef as well

Comment (ashwinr) [inline test/rules/noStringLiteralRuleTests.ts]: i disagree. it's unnecessarily verbose. please make it one line.

**Comment by adidahiya** [comment]
@ashwinr ok, updated

**Pull request merged**